From 3e950821df648e3d3e0f2b1bc190478b1e57b0b7 Mon Sep 17 00:00:00 2001
From: Yuan Juntao <juntao.yuan@intel.com>
Date: Wed, 10 Dec 2014 15:49:13 +0800
Subject: [PATCH] regulator: add regulators for sd with dollar cove pmic

Add vmmc & vqmmc regulators for SD host controller with
dollar cove pmic.

Change-Id: Ic99714619ada7fa66e6145b4ef5317868d5dc8d8
Signed-off-by: Yuan Juntao <juntao.yuan@intel.com>
---
 .../device_libs/platform_sdio_regulator.c          | 55 +++++++++++++++++++---
 drivers/mfd/intel_soc_pmic_dc.c                    | 13 +++++
 2 files changed, 61 insertions(+), 7 deletions(-)

Index: linux/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c
===================================================================
--- linux.orig/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c	2015-02-18 10:41:48.339850315 -0800
+++ linux/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c	2015-02-18 10:48:12.534821679 -0800
@@ -25,6 +25,7 @@
 #include <linux/platform_device.h>
 #include <asm/cpu_device_id.h>
 #include <linux/regulator/intel_whiskey_cove_pmic.h>
+#include <linux/regulator/intel_dollar_cove_pmic.h>
 
 struct acpi_ids { char *hid; char *uid; };
 
@@ -70,8 +71,10 @@
 }
 
 /* vsdcard regulator */
-static struct regulator_init_data ccove_vsdcard_data = {
+static struct regulator_init_data vmmc_data = {
 	.constraints = {
+		.min_uV = 3300000,
+		.max_uV = 3300000,
 		.valid_ops_mask		= REGULATOR_CHANGE_STATUS,
 		.valid_modes_mask	= REGULATOR_MODE_NORMAL,
 	},
@@ -82,13 +85,13 @@
 static struct fixed_voltage_config ccove_vsdcard = {
 	.supply_name	= "gpio_vsdcard",
 	.microvolts	= 3300000,
-	.init_data	= &ccove_vsdcard_data,
+	.init_data	= &vmmc_data,
 };
 
 /* vsdio regulator */
-static struct regulator_init_data ccove_vsdio_data = {
+static struct regulator_init_data vqmmc_data = {
 	.constraints = {
-		.min_uV			= 1700000,
+		.min_uV			= 1800000,
 		.max_uV			= 3300000,
 		.valid_ops_mask		= REGULATOR_CHANGE_VOLTAGE |
 					REGULATOR_CHANGE_STATUS,
@@ -122,10 +125,10 @@
 	.states		= ccove_vsdio_states,
 	.nr_states	= ARRAY_SIZE(ccove_vsdio_states),
 	.type		= REGULATOR_VOLTAGE,
-	.init_data	= &ccove_vsdio_data,
+	.init_data	= &vqmmc_data,
 };
 
-static void intel_setup_ccove_sd_regulators(void)
+static void intel_setup_crystal_cove_sd_regulators(void)
 {
 	struct device *dev = NULL;
 	struct acpi_ids *sdhc_ids;
@@ -205,11 +208,47 @@
 		sizeof(struct wcove_regulator_info), WCOVE_ID_VSDIO + 1);
 }
 
-static int __init sdio_regulator_init(void)
+/*************************************************************
+*
+* DCOVE SD card related regulator
+*
+*************************************************************/
+static struct regulator_init_data dcovex_vmmc_data;
+static struct regulator_init_data dcovex_vqmmc_data;
+
+static struct dcovex_regulator_info dcove_vmmc_info = {
+	.init_data = &dcovex_vmmc_data,
+};
+
+static struct dcovex_regulator_info dcove_vqmmc_info = {
+	.init_data = &dcovex_vqmmc_data,
+};
+
+static void intel_setup_dollar_cove_sd_regulators(void)
 {
-	intel_setup_ccove_sd_regulators();
+	memcpy((void *)&dcovex_vmmc_data, (void *)&vmmc_data,
+				sizeof(struct regulator_init_data));
+	memcpy((void *)&dcovex_vqmmc_data, (void *)&vqmmc_data,
+				sizeof(struct regulator_init_data));
+
+	/* set enable time for vqmmc regulator, stabilize power rail */
+	dcovex_vqmmc_data.constraints.enable_time = 10000;
+
+	dcovex_vmmc_data.constraints.name = "LDO_2";
+	dcovex_vqmmc_data.constraints.name = "LDO_3";
 
+	/* register SD card regulator for dollar cove PMIC */
+	intel_soc_pmic_set_pdata("dcovex_regulator", &dcove_vmmc_info,
+		sizeof(struct dcovex_regulator_info), DCOVEX_ID_LDO2 + 1);
+	intel_soc_pmic_set_pdata("dcovex_regulator", &dcove_vqmmc_info,
+		sizeof(struct dcovex_regulator_info), DCOVEX_ID_LDO3 + 1);
+}
+
+static int __init sdio_regulator_init(void)
+{
+	intel_setup_crystal_cove_sd_regulators();
 	intel_setup_whiskey_cove_sd_regulators();
+	intel_setup_dollar_cove_sd_regulators();
 
 	return 0;
 }
Index: linux/drivers/mfd/intel_soc_pmic_dc.c
===================================================================
--- linux.orig/drivers/mfd/intel_soc_pmic_dc.c	2015-02-18 10:41:48.339850315 -0800
+++ linux/drivers/mfd/intel_soc_pmic_dc.c	2015-02-18 10:41:48.335850284 -0800
@@ -25,6 +25,7 @@
 #include <linux/version.h>
 #include <linux/power/dc_xpwr_battery.h>
 #include <linux/power/dc_xpwr_charger.h>
+#include <linux/regulator/intel_dollar_cove_pmic.h>
 
 #include <asm/intel_em_config.h>
 #include <linux/extcon/extcon-dc-pwrsrc.h>
@@ -250,6 +251,18 @@
 	{
 		.name = "dollar_cove_region",
 	},
+	{
+		.name = "dcovex_regulator",
+		.id = DCOVEX_ID_LDO2 + 1,
+		.num_resources = 0,
+		.resources = NULL,
+	},
+	{
+		.name = "dcovex_regulator",
+		.id = DCOVEX_ID_LDO3 + 1,
+		.num_resources = 0,
+		.resources = NULL,
+	},
 	{NULL, },
 };
 
